### ISSUE TITLE
invoice.save returns false but invoice.errors is empty

### ISSUE THREAD

**Issue by virajvchavan** [waynerobinson/xeroizer#538]
> `invoice.save` returns false but `invoice.errors` is empty

`invoice` object is built with some params
```
invoice = xero.Invoice.build(...)
invoice.valid? # returns true
invoice.save # returns false
invoice.errors # returns empty array []
invoice.save! # raises Xeroizer::ApiException

```
How do I know what caused `invoice.save` to return false?

**Comment by virajvchavan** [comment]
Update: found the specific error by adding a logger and seeing the actual HTTP response. The `item_code` I sent for the `line_items` was incorrect.
Such obvious errors should be handled/shown in the code itself instead of raising a generic `Xeroizer::ApiException` exception.

**Comment by toxaq** [comment]
I have the same problem. Upgrading to 3.00 has broken validation errors on all models.

**Comment by toxaq** [comment]
It seems in version 2.18 the save! (bang) method was introduced. Previously validation errors would throw an exception. There is no code, that I could find, that will populate the errors from the validation errors in the XML response.

If you use the bang method (save!) then you can catch the original validation exception and handle this yourself.

**Comment by rjaus** [comment]
I'm not seeing this behaviour.

```
3.0.0 :034 > invoice = client.Invoice.build
 => #<Xeroizer::Record::Invoice > 
3.0.0 :035 > invoice.valid?
 => false 
3.0.0 :036 > invoice.save
 => false 
3.0.0 :037 > invoice.errors
 => [[:date, "can't be blank"], [:type, "not one of ACCPAY, ACCREC"], [:contact, "must be valid"]] 
3.0.0 :038 > invoice.save!
Traceback (most recent call last):
        5: from /Users/riley.james/.rvm/rubies/ruby-3.0.0/bin/irb:23:in `<main>'
        4: from /Users/riley.james/.rvm/rubies/ruby-3.0.0/bin/irb:23:in `load'
        3: from /Users/riley.james/.rvm/rubies/ruby-3.0.0/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        2: from (irb):38:in `<main>'
        1: from /Users/riley.james/Code/xeroizer/lib/xeroizer/record/base.rb:116:in `save!'
Xeroizer::RecordInvalid (Xeroizer::RecordInvalid)

```
![Screen Shot 2021-09-23 at 12 05 31 PM](https://user-images.githubusercontent.com/773626/134437840-a004d3f4-ac26-4fbe-bdc3-fa5d80cb724f.png)

Did someone fix it?

Feel free to re-open if this issue persists, but I'm unable to reproduce based on the above details provided.

**Comment by toxaq** [comment]
Those are all local validation errors. From memory, this was about errors returned from the API not been populated anywhere. So if the contact already exists for example, the response is just captured and disregarded.

**Comment by rjaus** [comment]
Right, thanks for details.

Yes, I've also noticed that OAuth related errors such as token invalid / expired, return false, but it's unclear unless you're catching those errors from save!

**Comment by toxaq** [comment]
Yes, our original logic utilised the nature of save throwing errors so it was simple for us to switch to the bang method (after the 2.18 change) to get back to where we were. Ideally though the save method would do as OP expected.

**Comment by rjaus** [comment]
Ok, seems like a relatively easy fix.  None of the error classes were included in the OAuth2 class.

I'll get a PR up shortly, @toxaq I'll ping you to review, be good to have some eyes on it as I'm a bit dubious on the test coverage.

**Comment by rjaus** [comment]
Looks like things changed when this commit came in:
https://github.com/waynerobinson/xeroizer/commit/3f7f1739f6a168816a8eb1693530357812fdf032
https://github.com/waynerobinson/xeroizer/commit/4a181bc47f4663dc830e861f1121141381b2400b

But it's not super clear to me where to take it to improve.

**Comment by toxaq** [comment]
I would probably start around here:
https://github.com/waynerobinson/xeroizer/blob/b78a9a0b1933649b74cb3877c56b6b702ac0cc33/lib/xeroizer/record/base.rb#L110

The exceptions are still be raised, as per the pre-2.18 functionality, they're jus captured and dumped.

**Comment by toxaq** [comment]
Could be as simple as

    self.errors[:base] << e.message

There's nothing else available error wise.

**Comment by joseairosa** [comment]
Having the same issue here, any updates?
